### PR TITLE
Fix non-determinism in Regex source generator

### LINKED ISSUE #76774: build libs with `/p:Deterministic=true`  some libraray make diff
I build libs with /p:Deterministic=true with the below command:
> ./build.cmd -pack -runtimeFlavor mono /p:Deterministic=true 

when I rebuild some libraries generate diff:
image

Is there tool to check they are the same code assembly?

### PR DESCRIPTION
**Pull request by stephentoub**
The source generator enumerates a Hashtable to write out its contents.  When the keys of the Hashtable are strings, string hash code randomization may result in the order of that enumeration being different in different processes, leading to non-deterministic ordering of values written out and thus non-deterministic source generator output.

Fixes https://github.com/dotnet/runtime/issues/76774 (though we'll want to backport this to release/7.0)
cc: @MichalStrehovsky, @ViktorHofer, @joperezr 

@jaredpar, another source of non-determinism for your list.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs [hunk 1]
```diff
@@ -123,13 +123,13 @@ private static void EmitRegexPartialMethod(RegexMethod regexMethod, IndentedText
             if (rm.Tree.CaptureNumberSparseMapping is not null)
             {
                 writer.Write("        base.Caps = new Hashtable {");
-                AppendHashtableContents(writer, rm.Tree.CaptureNumberSparseMapping);
+                AppendHashtableContents(writer, rm.Tree.CaptureNumberSparseMapping.Cast<DictionaryEntry>());
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:126]: This Hashtable has Int32s as keys, so there's no randomization involved.  I could add one for good measure or if we just like the sorted output.

[the remark this answers is not in the document]
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs:126]: Seems harmless and maybe a bit clearer.

**Review by joperezr — APPROVED** [review]
LGTM

**Pull request merged**